File: bindings/dart/test/e2e_test.dart
import 'package:test/test.dart';
import 'package:strling/strling.dart';

/// E2E Tests - Black-box testing where DSL input produces a regex
/// that matches expected strings.
void main() {
  group('E2E Phone Number', () {
    test('matches valid formats', () {
      final regex = compileToRegex(r'^(\d{3})[-. ]?(\d{3})[-. ]?(\d{4})$');
      final re = RegExp(regex);

      expect(re.hasMatch('555-123-4567'), isTrue);
      expect(re.hasMatch('555.123.4567'), isTrue);
      expect(re.hasMatch('555 123 4567'), isTrue);
      expect(re.hasMatch('5551234567'), isTrue);
    });

    test('rejects invalid formats', () {
      final regex = compileToRegex(r'^(\d{3})[-. ]?(\d{3})[-. ]?(\d{4})$');
      final re = RegExp(regex);

      expect(re.hasMatch('55-123-4567'), isFalse);
      expect(re.hasMatch('555-12-4567'), isFalse);
      expect(re.hasMatch('555-123-456'), isFalse);
      expect(re.hasMatch('abc-def-ghij'), isFalse);
    });
  });

  group('E2E Email', () {
    test('matches valid formats', () {
      final regex = compileToRegex(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$');
      final re = RegExp(regex);

      expect(re.hasMatch('user@example.com'), isTrue);
      expect(re.hasMatch('test.user@domain.org'), isTrue);
    });

    test('rejects invalid formats', () {
      final regex = compileToRegex(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$');
      final re = RegExp(regex);

      expect(re.hasMatch('@example.com'), isFalse);
      expect(re.hasMatch('user@'), isFalse);
      expect(re.hasMatch('user@.com'), isFalse);
    });
  });

  group('E2E IPv4', () {
    test('matches valid addresses', () {
      final regex = compileToRegex(r'^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$');
      final re = RegExp(regex);

      expect(re.hasMatch('192.168.1.1'), isTrue);
      expect(re.hasMatch('10.0.0.1'), isTrue);
      expect(re.hasMatch('255.255.255.255'), isTrue);
      expect(re.hasMatch('0.0.0.0'), isTrue);
    });

    test('rejects invalid addresses', () {
      final regex = compileToRegex(r'^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$');
      final re = RegExp(regex);

      expect(re.hasMatch('192.168.1'), isFalse);
      expect(re.hasMatch('192.168.1.1.1'), isFalse);
      expect(re.hasMatch('192-168-1-1'), isFalse);
    });
  });

  group('E2E Hex Color', () {
    test('matches valid colors', () {
      final regex = compileToRegex(r'^#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})$');
      final re = RegExp(regex);

      expect(re.hasMatch('#ffffff'), isTrue);
      expect(re.hasMatch('#000000'), isTrue);
      expect(re.hasMatch('#ABC123'), isTrue);
      expect(re.hasMatch('#fff'), isTrue);
      expect(re.hasMatch('#F00'), isTrue);
    });

    test('rejects invalid colors', () {
      final regex = compileToRegex(r'^#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})$');
      final re = RegExp(regex);

      expect(re.hasMatch('ffffff'), isFalse);
      expect(re.hasMatch('#ffff'), isFalse);
      expect(re.hasMatch('#GGGGGG'), isFalse);
    });
  });

  group('E2E Date', () {
    test('matches valid dates', () {
      final regex = compileToRegex(r'^(\d{4})-(\d{2})-(\d{2})$');
      final re = RegExp(regex);

      expect(re.hasMatch('2024-01-15'), isTrue);
      expect(re.hasMatch('2000-12-31'), isTrue);
      expect(re.hasMatch('1999-06-30'), isTrue);
    });

    test('rejects invalid dates', () {
      final regex = compileToRegex(r'^(\d{4})-(\d{2})-(\d{2})$');
      final re = RegExp(regex);

      expect(re.hasMatch('24-01-15'), isFalse);
      expect(re.hasMatch('2024/01/15'), isFalse);
      expect(re.hasMatch('2024-1-15'), isFalse);
    });
  });

  group('E2E Lookahead', () {
    test('positive lookahead', () {
      final regex = compileToRegex('foo(?=bar)');
      final re = RegExp(regex);

      expect(re.hasMatch('foobar'), isTrue);
      expect(re.hasMatch('foobaz'), isFalse);
    });

    test('negative lookahead', () {
      final regex = compileToRegex('foo(?!bar)');
      final re = RegExp(regex);

      expect(re.hasMatch('foobaz'), isTrue);
    });
  });

  group('E2E Word Boundary', () {
    test('matches word boundaries', () {
      final regex = compileToRegex(r'\bword\b');
      final re = RegExp(regex);

      expect(re.hasMatch('word'), isTrue);
      expect(re.hasMatch('a word here'), isTrue);
      expect(re.hasMatch('sword'), isFalse);
      expect(re.hasMatch('wording'), isFalse);
    });
  });

  group('E2E Alternation', () {
    test('matches alternatives', () {
      final regex = compileToRegex(r'^(cat|dog|bird)$');
      final re = RegExp(regex);

      expect(re.hasMatch('cat'), isTrue);
      expect(re.hasMatch('dog'), isTrue);
      expect(re.hasMatch('bird'), isTrue);
      expect(re.hasMatch('cats'), isFalse);
      expect(re.hasMatch('fish'), isFalse);
    });
  });

  group('E2E Quantifiers', () {
    test('plus', () {
      final regex = compileToRegex(r'^a+$');
      final re = RegExp(regex);

      expect(re.hasMatch('a'), isTrue);
      expect(re.hasMatch('aa'), isTrue);
      expect(re.hasMatch('aaa'), isTrue);
      expect(re.hasMatch(''), isFalse);
      expect(re.hasMatch('b'), isFalse);
    });

    test('star', () {
      final regex = compileToRegex(r'^a*$');
      final re = RegExp(regex);

      expect(re.hasMatch(''), isTrue);
      expect(re.hasMatch('a'), isTrue);
      expect(re.hasMatch('aaa'), isTrue);
      expect(re.hasMatch('b'), isFalse);
    });

    test('optional', () {
      final regex = compileToRegex(r'^a?$');
      final re = RegExp(regex);

      expect(re.hasMatch(''), isTrue);
      expect(re.hasMatch('a'), isTrue);
      expect(re.hasMatch('aa'), isFalse);
    });

    test('exact', () {
      final regex = compileToRegex(r'^a{3}$');
      final re = RegExp(regex);

      expect(re.hasMatch('aaa'), isTrue);
      expect(re.hasMatch('a'), isFalse);
      expect(re.hasMatch('aa'), isFalse);
      expect(re.hasMatch('aaaa'), isFalse);
    });

    test('range', () {
      final regex = compileToRegex(r'^a{2,4}$');
      final re = RegExp(regex);

      expect(re.hasMatch('aa'), isTrue);
      expect(re.hasMatch('aaa'), isTrue);
      expect(re.hasMatch('aaaa'), isTrue);
      expect(re.hasMatch('a'), isFalse);
      expect(re.hasMatch('aaaaa'), isFalse);
    });

    test('atLeast', () {
      final regex = compileToRegex(r'^a{2,}$');
      final re = RegExp(regex);

      expect(re.hasMatch('aa'), isTrue);
      expect(re.hasMatch('aaa'), isTrue);
      expect(re.hasMatch(''), isFalse);
      expect(re.hasMatch('a'), isFalse);
    });
  });

  group('E2E Capture Groups', () {
    test('extracts captured groups', () {
      final regex = compileToRegex(r'^(\d{4})-(\d{2})-(\d{2})$');
      final re = RegExp(regex);
      final match = re.firstMatch('2024-12-25');

      expect(match, isNotNull);
      expect(match!.group(1), equals('2024'));
      expect(match.group(2), equals('12'));
      expect(match.group(3), equals('25'));
    });
  });
}

String compileToRegex(String dsl) {
  final result = Parser.parse(dsl);
  final ir = result.node.toIR();
  return Pcre2Emitter.emit(ir, result.flags);
}

File: bindings/kotlin/src/test/kotlin/strling/E2ETest.kt
package strling

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.assertFalse
import strling.core.Parser
import strling.core.Compiler
import strling.emitters.Pcre2Emitter

/**
 * E2E Tests - Black-box testing where DSL input produces a regex
 * that matches expected strings.
 */
class E2ETest {

    private fun compileToRegex(dsl: String): String {
        val (flags, ast) = Parser.parse(dsl)
        val ir = Compiler.compile(ast)
        return Pcre2Emitter.emit(ir, flags)
    }

    // ========================================================================
    // Phone Number E2E Tests
    // ========================================================================

    @Test
    fun `E2E Phone Number - matches valid formats`() {
        val regex = compileToRegex("^(\\d{3})[-. ]?(\\d{3})[-. ]?(\\d{4})\$")
        val re = Regex(regex)

        assertTrue(re.matches("555-123-4567"))
        assertTrue(re.matches("555.123.4567"))
        assertTrue(re.matches("555 123 4567"))
        assertTrue(re.matches("5551234567"))
    }

    @Test
    fun `E2E Phone Number - rejects invalid formats`() {
        val regex = compileToRegex("^(\\d{3})[-. ]?(\\d{3})[-. ]?(\\d{4})\$")
        val re = Regex(regex)

        assertFalse(re.matches("55-123-4567"))
        assertFalse(re.matches("555-12-4567"))
        assertFalse(re.matches("555-123-456"))
        assertFalse(re.matches("abc-def-ghij"))
    }

    // ========================================================================
    // Email E2E Tests
    // ========================================================================

    @Test
    fun `E2E Email - matches valid formats`() {
        val regex = compileToRegex("^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}\$")
        val re = Regex(regex)

        assertTrue(re.matches("user@example.com"))
        assertTrue(re.matches("test.user@domain.org"))
    }

    @Test
    fun `E2E Email - rejects invalid formats`() {
        val regex = compileToRegex("^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}\$")
        val re = Regex(regex)

        assertFalse(re.matches("@example.com"))
        assertFalse(re.matches("user@"))
        assertFalse(re.matches("user@.com"))
    }

    // ========================================================================
    // IPv4 E2E Tests
    // ========================================================================

    @Test
    fun `E2E IPv4 - matches valid addresses`() {
        val regex = compileToRegex("^(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\$")
        val re = Regex(regex)

        assertTrue(re.matches("192.168.1.1"))
        assertTrue(re.matches("10.0.0.1"))
        assertTrue(re.matches("255.255.255.255"))
        assertTrue(re.matches("0.0.0.0"))
    }

    @Test
    fun `E2E IPv4 - rejects invalid addresses`() {
        val regex = compileToRegex("^(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\$")
        val re = Regex(regex)

        assertFalse(re.matches("192.168.1"))
        assertFalse(re.matches("192.168.1.1.1"))
        assertFalse(re.matches("192-168-1-1"))
    }

    // ========================================================================
    // Hex Color E2E Tests
    // ========================================================================

    @Test
    fun `E2E Hex Color - matches valid colors`() {
        val regex = compileToRegex("^#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})\$")
        val re = Regex(regex)

        assertTrue(re.matches("#ffffff"))
        assertTrue(re.matches("#000000"))
        assertTrue(re.matches("#ABC123"))
        assertTrue(re.matches("#fff"))
        assertTrue(re.matches("#F00"))
    }

    @Test
    fun `E2E Hex Color - rejects invalid colors`() {
        val regex = compileToRegex("^#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})\$")
        val re = Regex(regex)

        assertFalse(re.matches("ffffff"))
        assertFalse(re.matches("#ffff"))
        assertFalse(re.matches("#GGGGGG"))
    }

    // ========================================================================
    // Date E2E Tests
    // ========================================================================

    @Test
    fun `E2E Date - matches valid dates`() {
        val regex = compileToRegex("^(\\d{4})-(\\d{2})-(\\d{2})\$")
        val re = Regex(regex)

        assertTrue(re.matches("2024-01-15"))
        assertTrue(re.matches("2000-12-31"))
        assertTrue(re.matches("1999-06-30"))
    }

    @Test
    fun `E2E Date - rejects invalid dates`() {
        val regex = compileToRegex("^(\\d{4})-(\\d{2})-(\\d{2})\$")
        val re = Regex(regex)

        assertFalse(re.matches("24-01-15"))
        assertFalse(re.matches("2024/01/15"))
        assertFalse(re.matches("2024-1-15"))
    }

    // ========================================================================
    // Lookahead E2E Tests
    // ========================================================================

    @Test
    fun `E2E Positive Lookahead`() {
        val regex = compileToRegex("foo(?=bar)")
        val re = Regex(regex)

        assertTrue(re.containsMatchIn("foobar"))
        assertFalse(re.containsMatchIn("foobaz"))
    }

    @Test
    fun `E2E Negative Lookahead`() {
        val regex = compileToRegex("foo(?!bar)")
        val re = Regex(regex)

        assertTrue(re.containsMatchIn("foobaz"))
    }

    // ========================================================================
    // Word Boundary E2E Tests
    // ========================================================================

    @Test
    fun `E2E Word Boundary`() {
        val regex = compileToRegex("\\bword\\b")
        val re = Regex(regex)

        assertTrue(re.containsMatchIn("word"))
        assertTrue(re.containsMatchIn("a word here"))
        assertFalse(re.containsMatchIn("sword"))
        assertFalse(re.containsMatchIn("wording"))
    }

    // ========================================================================
    // Alternation E2E Tests
    // ========================================================================

    @Test
    fun `E2E Alternation`() {
        val regex = compileToRegex("^(cat|dog|bird)\$")
        val re = Regex(regex)

        assertTrue(re.matches("cat"))
        assertTrue(re.matches("dog"))
        assertTrue(re.matches("bird"))
        assertFalse(re.matches("cats"))
        assertFalse(re.matches("fish"))
    }

    // ========================================================================
    // Quantifier E2E Tests
    // ========================================================================

    @Test
    fun `E2E Quantifier Plus`() {
        val regex = compileToRegex("^a+\$")
        val re = Regex(regex)

        assertTrue(re.matches("a"))
        assertTrue(re.matches("aa"))
        assertTrue(re.matches("aaa"))
        assertFalse(re.matches(""))
        assertFalse(re.matches("b"))
    }

    @Test
    fun `E2E Quantifier Star`() {
        val regex = compileToRegex("^a*\$")
        val re = Regex(regex)

        assertTrue(re.matches(""))
        assertTrue(re.matches("a"))
        assertTrue(re.matches("aaa"))
        assertFalse(re.matches("b"))
    }

    @Test
    fun `E2E Quantifier Optional`() {
        val regex = compileToRegex("^a?\$")
        val re = Regex(regex)

        assertTrue(re.matches(""))
        assertTrue(re.matches("a"))
        assertFalse(re.matches("aa"))
    }

    @Test
    fun `E2E Quantifier Exact`() {
        val regex = compileToRegex("^a{3}\$")
        val re = Regex(regex)

        assertTrue(re.matches("aaa"))
        assertFalse(re.matches("a"))
        assertFalse(re.matches("aa"))
        assertFalse(re.matches("aaaa"))
    }

    @Test
    fun `E2E Quantifier Range`() {
        val regex = compileToRegex("^a{2,4}\$")
        val re = Regex(regex)

        assertTrue(re.matches("aa"))
        assertTrue(re.matches("aaa"))
        assertTrue(re.matches("aaaa"))
        assertFalse(re.matches("a"))
        assertFalse(re.matches("aaaaa"))
    }

    @Test
    fun `E2E Quantifier AtLeast`() {
        val regex = compileToRegex("^a{2,}\$")
        val re = Regex(regex)

        assertTrue(re.matches("aa"))
        assertTrue(re.matches("aaa"))
        assertTrue(re.matches("aaaa"))
        assertFalse(re.matches(""))
        assertFalse(re.matches("a"))
    }

    // ========================================================================
    // Capture Groups E2E Tests
    // ========================================================================

    @Test
    fun `E2E Capture Groups`() {
        val regex = compileToRegex("^(\\d{4})-(\\d{2})-(\\d{2})\$")
        val re = Regex(regex)
        val match = re.matchEntire("2024-12-25")

        assertEquals("2024", match?.groupValues?.get(1))
        assertEquals("12", match?.groupValues?.get(2))
        assertEquals("25", match?.groupValues?.get(3))
    }
}
